kingpin/actors/aws/s3.py

```diff
@@ -201,16 +235,94 @@ def _delete_bucket(self, bucket):
+
+        (Note, this method is longer than we'd like .. but in this Bucket actor
+        is going to do _a lot_ of things, so encapsulating the logic all in a
+        single method makes the rest of the code easier to read and
+        understand.)
+
+        args:
+            bucket: The S3 bucket object as returned by Boto
+        """
+        new = self.policy
+        exist = {}
+
+        # Get our existing policy and convert it into a dict we can deal with
+        try:
+            raw = yield self.thread(bucket.get_policy)
+            exist = json.loads(raw)
+        except S3ResponseError as e:
+            if e.status != 404:
+                raise exceptions.RecoverableActorFailure(
+                    'An unexpected error occurred: %s' % e)
+
+        # Now, if we're deleting the policy (policy=''), then optionally do
+        # that and bail.
+        if new == '':
+            # If no existing policy was found, just get us out of this function
+            if not exist:
+                raise gen.Return()
+
+            if self._dry:
+                self.log.warning('Would have deleted bucket policy')
+                raise gen.Return()
+
+            self.log.info('Deleting bucket policy')
+            try:
+                yield self.thread(bucket.delete_policy)
+            except S3ResponseError as e:
+                raise exceptions.RecoverableActorFailure(
+                    'An unexpected error occurred: %s' % e)
+            raise gen.Return()
+
+        # Now, diff our new policy from the existing policy. If there is no
+        # difference, then we bail out of the method.
+        diff = self._diff_policy_json(exist, new)
+        if not diff:
+            self.log.debug('Bucket policy matches')
+            raise gen.Return()
+
+        # Now, print out the diff..
+        self.log.info('Bucket policy differs from Amazons:')
+        for line in diff.split('\n'):
+            self.log.info('Diff: %s' % line)
+
+        # Final DRY check ... if dry, get out of here.
+        if self._dry:
+            self.log.warning('Would have pushed bucket policy %s'
+                             % self.option('policy'))
+            raise gen.Return()
+
+        # Push the new policy!
+        self.log.info('Pushing bucket policy %s' % self.option('policy'))
+        self.log.debug('Policy doc: %s' % new)
+        try:
+            yield self.thread(bucket.set_policy, json.dumps(new))
+        except S3ResponseError as e:
+            if e.error_code == 'MalformedPolicy':
+                raise base.InvalidPolicy(e.message)
+
+            raise exceptions.RecoverableActorFailure(
+                'An unexpected error occurred: %s' % e)
+        raise gen.Return()
+
+    @gen.coroutine
     def _execute(self):
         """Executes an actor and yields the results when its finished.
 
         raises: gen.Return(True)
         """
-        yield self._ensure_bucket()
+        bucket = yield self._ensure_bucket()
```

Comment:
Kind of a weird pattern here. Half passed-variables, half class-variables. This entire _execute could be literally 2 lines:

```
yield self._ensure_bucket()
yield self._ensure_policy()
```

Just like `self.policy` is stored you can can store self.bucket which you calculate _ensure_bucket. The docstring, as well as the comments and the final raise return are pretty useless